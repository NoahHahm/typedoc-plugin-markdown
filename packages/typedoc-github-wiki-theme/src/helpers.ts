export function parseUrl(url: string) {
  return encodeURI('../wiki/' + url.replace('.md', ''));
}
