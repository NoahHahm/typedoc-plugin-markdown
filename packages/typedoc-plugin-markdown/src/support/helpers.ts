import {
  ContainerReflection,
  DeclarationReflection,
  ProjectReflection,
  ReflectionKind,
  SignatureReflection,
} from 'typedoc';
import { backTicks } from './els';

/**
 * Determines if current signature is a constructor
 */
export const isConstructor = (
  signature: DeclarationReflection | SignatureReflection,
) => {
  return signature.parent?.kindOf(ReflectionKind.Constructor);
};

export function getIndexHeadingLevel(
  reflection: DeclarationReflection | ProjectReflection,
) {
  if (
    reflection.kindOf([
      ReflectionKind.Project,
      ReflectionKind.Module,
      ReflectionKind.Namespace,
    ]) ||
    reflection.hasOwnDocument
  ) {
    return 2;
  }
  return 4;
}

export function getGroupHeadingLevel(container: ContainerReflection) {
  return container.hasOwnDocument ? 2 : 4;
}

export function getReflectionHeadingLevel(
  reflection: DeclarationReflection | SignatureReflection,
) {
  if (reflection.hasOwnDocument) {
    return 1;
  }
  if (reflection.kindOf(ReflectionKind.Constructor)) {
    return reflection.parent?.hasOwnDocument ? 2 : 4;
  }
  return reflection.parent?.hasOwnDocument ? 3 : 5;
}

export function getMemberHeadingLevel(
  reflection: DeclarationReflection | SignatureReflection,
) {
  return getReflectionHeadingLevel(reflection) + 1;
}

export function getReflectionTitle(reflection: DeclarationReflection) {
  const md: string[] = [];

  md.push(reflection.name);

  if (reflection.typeParameters) {
    const typeParameters = reflection.typeParameters
      .map((typeParameter) => typeParameter.name)
      .join(', ');
    md.push(`\\<${typeParameters}\\>`);
  }

  return md.join('');
}

export function getFlags(reflection: DeclarationReflection) {
  if (reflection.flags?.length > 0 && !reflection.flags.isRest) {
    return reflection.flags.map((flag) => backTicks(flag)).join(' ');
  }
  return null;
}
