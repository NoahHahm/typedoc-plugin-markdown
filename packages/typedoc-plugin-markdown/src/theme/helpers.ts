/**
 *
 * A set of helpers that parses TypeDoc models to be consumed by theme resources.
 *
 * @module
 */

import {
  DeclarationReflection,
  ProjectReflection,
  ReflectionKind,
} from 'typedoc';
import { backTicks } from '../support/elements';
import { escapeChars } from '../support/utils';

export function getDeclarationType(declaration: DeclarationReflection) {
  if (declaration.getSignature) {
    return declaration.getSignature.type;
  }
  if (declaration.setSignature) {
    return declaration.setSignature.type;
  }
  return declaration.type;
}

export function getProjectDisplayName(
  project: ProjectReflection,
  includeVersion: boolean,
): string {
  const version =
    includeVersion && project.packageVersion
      ? ` - v${project.packageVersion}`
      : '';
  return `${project.name}${version ? version : ''}`;
}

export function hasIndex(
  reflection: DeclarationReflection | ProjectReflection,
) {
  return reflection.groups?.some((group) => group.allChildrenHaveOwnDocument());
}

export function hasTOC(reflection: DeclarationReflection | ProjectReflection) {
  return (
    reflection.kindOf([
      ReflectionKind.Project,
      ReflectionKind.Module,
      ReflectionKind.Namespace,
    ]) &&
    reflection.groups?.some((group) => !group.allChildrenHaveOwnDocument())
  );
}

export function isGroupKind(reflection: DeclarationReflection) {
  return reflection.kindOf([
    ReflectionKind.Class,
    ReflectionKind.Interface,
    ReflectionKind.Enum,
    ReflectionKind.Function,
    ReflectionKind.Variable,
    ReflectionKind.TypeAlias,
  ]);
}

export function getModifier(reflection: DeclarationReflection) {
  if (reflection.flags.isAbstract) {
    return 'abstract';
  }
  if (reflection.flags.isPrivate) {
    return 'private';
  }
  if (reflection.flags.isReadonly) {
    return 'readonly';
  }
  if (reflection.flags.isStatic) {
    return 'static';
  }
  if (reflection.getSignature) {
    return 'get';
  }
  if (reflection.setSignature) {
    return 'set';
  }
  return null;
}

export const KEYWORD_MAP = {
  [ReflectionKind.Class]: 'class',
  [ReflectionKind.Interface]: 'interface',
  [ReflectionKind.Enum]: 'enum',
  [ReflectionKind.TypeAlias]: 'type',
};

export function getMemberTitle(reflection: DeclarationReflection) {
  const md: string[] = [];

  const name: string[] = [];

  const flagsToInclude = ['Abstract'];

  const flags = reflection.flags.filter((flag) =>
    flagsToInclude.includes(flag),
  );

  if (flags.length) {
    name.push(flags.map((flag) => backTicks(flag.toLowerCase())).join(' '));
    name.push(' ');
  }

  name.push(`${escapeChars(reflection.name)}`);

  if (reflection.typeParameters) {
    const typeParameters = reflection.typeParameters
      .map((typeParameter) => typeParameter.name)
      .join(', ');
    name.push(`${backTicks(`<${typeParameters}>`)}`);
  }

  md.push(name.join(''));

  return md.join(': ');
}
