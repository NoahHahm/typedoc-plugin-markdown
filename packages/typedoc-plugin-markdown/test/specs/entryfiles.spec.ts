import {
  FixtureOutputDir,
  FixtureOutputFileStrategy,
} from '../__utils__/fixture-config';
import { expectFileToEqual } from '../__utils__/helpers';

describe(`EntryFiles`, () => {
  test(`should get entry module`, () => {
    expectFileToEqual(
      FixtureOutputDir.EntryFiles,
      FixtureOutputFileStrategy.Members,
      ['README.md', 'index.md'],
    );
  });

  test(`should get readme file`, () => {
    expectFileToEqual(
      FixtureOutputDir.EntryFiles,
      FixtureOutputFileStrategy.Members,
      'readme_.md',
      1,
    );
  });
});
