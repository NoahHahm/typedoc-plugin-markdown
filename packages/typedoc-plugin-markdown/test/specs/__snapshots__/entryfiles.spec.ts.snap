// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EntryFiles should get entry module: (Output File Strategy "members") (Option Group "1") 1`] = `
"typedoc-stubs • [Readme](readme_.md) \\| [Documentation](docs.md)

***

# entry-module

## Index

### Classes

- [ModuleClass](entry-module/classes/ModuleClass.md)
"
`;

exports[`EntryFiles should get entry module: (Output File Strategy "members") (Option Group "2") 1`] = `
"typedoc-stubs • [Documentation](docs.md)

***

# entry-module

## Index

### Classes

- [ModuleClass](entry-module/classes/ModuleClass.md)
"
`;

exports[`EntryFiles should get readme file: (Output File Strategy "members") (Option Group "1") 1`] = `
"typedoc-stubs • [Readme](readme_.md) \\| [Documentation](docs.md)

***

# MY README

> Some block quote

Some text.

## Readme heading

Some text.
"
`;
